working_copy: existing submodule directory is ok. Fixes submodule history traversion issue.

thomasa88 opened the pull request:
Replaces https://github.com/jj-vcs/jj/pull/8176

Fixes #8065.

This commit fixes the problem of jj marking a git submodule as a deleted path,
after the user moves (edit/new) to a commit where the submodule did not exist
and then back to a commit where it existed. Failure to remove a
submodule's directory is no longer considered a conflict - if it was not going
to be replaced by another file.


<!--
There's no need to add anything here, but feel free to add a personal message.
Please describe the changes in this PR in the commit message(s) instead, with
each commit representing one logical change. Address code review comments by
rewriting the commits rather than adding commits on top. Use force-push when
pushing the updated commits (`jj git push` does that automatically when you
rewrite commits). Merge the PR at will once it's been approved. See
https://github.com/jj-vcs/jj/blob/main/docs/contributing.md for details.
Note that you need to sign Google's CLA to contribute.
-->

# Checklist

If applicable:

- [x] I have updated `CHANGELOG.md`
- N/A I have updated the documentation (`README.md`, `docs/`, `demos/`)
- N/A I have updated the config schema (`cli/src/config-schema.json`)
- [x] I have added/updated tests to cover my changes

thomasa88 commented on lib/src/local_working_copy.rs:
```diff
@@ -1967,21 +1992,45 @@ impl TreeState {
+                    eprintln!(
+                        "Files from the submodule will appear as added. Add the submodule to `.git/info/exclude` and run `jj file untrack` to ignore the directory contents."
+                    );
+                }
```
This `if` needs correct logic to check if the path is ignored. Or we just drop the print.

thomasa88 commented:
I dropped the commit paragraph about jj not automatically ignoring submodules, as it does since v0.36. Rebased on main and fixed conflicts.

thomasa88 commented:
Rebased on 0.40, then main.

martinvonz commented on lib/src/local_working_copy.rs:
```diff
@@ -744,6 +744,31 @@ fn remove_old_file(disk_path: &Path) -> Result<bool, CheckoutError> {
+/// untracked or modified files. This is in line with Git's behavior.
+///
+/// If the existing directory points to ".git" or ".jj", this function returns
+/// an error.
```
This would mean that there used to be a submodule called `.git` or `.jj`. If that was the case, it still seems fine to delete an empty directory by that name, or what do you think?

thomasa88 replied:
That's true! Check for reserved directories removed (and the function comment updated).

martinvonz commented on lib/src/local_working_copy.rs:
```diff
@@ -2309,6 +2340,11 @@ impl TreeState {
                     eprintln!("ignoring git submodule at {path:?}");
+                    // Git behavior: Create the submodule directory but don't
+                    // populate/overwrite the contents.
+                    if fs::create_dir_all(&disk_path).is_err() {
```
nit: `fs::create_dir()` instead of `create_dir_all()`? The parent directory should already exist at this point

thomasa88 replied:
Sneakily, the target directory might also already exist and `create_dir_all()` will ignore that it does, not leading to "failed to create submodule directory" being printed. That is, if the directory already exists, we don't warn about failing to create it.

This very much relates to the fall-through conversation in https://github.com/jj-vcs/jj/pull/8225/changes#r3068634734 .

thomasa88 replied:
I pondered a bit whether to skip the fall-throughs, but especially the parent deletion code did not feel nice to duplicate inside the current fall-through code.

I added a comment to motivate create_dir_all().

martinvonz replied:
I see. Could avoid the comment by making the code clearer like this then?

```rust
match fs::create_dir(&disk_path) {
    Ok(()) | Err(err) if err.kind() == ErrorKind::AlreadyExists => {},
    Err(_) => eprintln!("warning: failed to create submodule directory {path:?}"),
}
```

thomasa88 replied:
Good point. Done! (I had to split out the Ok due to the guard)

martinvonz commented on lib/src/local_working_copy.rs:
```diff
@@ -2309,6 +2340,11 @@ impl TreeState {
+                    // Git behavior: Create the submodule directory but don't
+                    // populate/overwrite the contents.
+                    if fs::create_dir_all(&disk_path).is_err() {
+                        eprintln!("warning: failed to create submodule directory {path:?}");
```
nit: Maybe better to propagate the error?

thomasa88 replied:
It does not change any state that jj cares about, so I think I'm inclined to let an error like this slip and let the user change to another commit even though an empty directory could not be created.

martinvonz commented on lib/src/local_working_copy.rs:
```diff
@@ -2294,9 +2293,21 @@ impl TreeState {
-                changed_file_states.push((path, FileState::placeholder()));
-                stats.skipped_files += 1;
-                return Ok(());
+                if disk_path.is_dir() && matches!(after, MaterializedTreeValue::GitSubmodule(_)) {
```
nit: Swap the operands so we do the cheap in-memory check on `after` before the more expensive system call for `is_dir()`

thomasa88 replied:
Done!

martinvonz commented on lib/src/local_working_copy.rs:
```diff
@@ -2294,9 +2293,21 @@ impl TreeState {
+                {
+                    // Failing to delete un-tracked submodule directory is ok,
+                    // as the, possibly untracked, contents would otherwise be
+                    // lost.
```
It wasn't obvious to me that falling through in these two cases was right, but it does seem to be. We want to delete parent directories if there was no submodule directory here, for example. (No action requested.)

thomasa88 replied:
I had to get this fully into context again to understand my reasoning.

I think that the case you mention is valid if `present_file_deleted = false` due to the directory not existing at all and a new test file (`can_create_new_file()`) cannot be created in its place ((Closest) parent directory not existing, or possibly hitting max path depth or the parent being an r/o directory).

The cases I mainly wanted to hit:

1. I want to hit `let file_state = match ...... FileState::for_gitsubmodule()` and `changed_file_states.push((path, file_state));` at the end of the lambda. I just saw that this is only relevant for a submodule coming into existence.

2. Also, for a submodule directory coming into existence, I want to hit `eprintln!("ignoring git submodule at {path:?}");` (about line 2349), to get info to the user.

I tried to improve the comments now. Also, I dropped the use of the word "ok", as it kept leading my thougths to `return Ok()`.

thomasa88 commented:
Thanks for the review! I implemented your last suggestion about create_dir/_all() and I also touched up the message of the last commit a bit, seeing it with fresh eyes.

Rebased on main. I will let it putter in my head a bit before doing the merge.
